[ExtendLifetimes] Implement llvm.fake.use to extend variable lifetimes

SLTozer opened the pull request:
This patch is part of a set of patches that add an `-fextend-lifetimes` flag to clang, which extends the lifetimes of local variables and parameters for improved debuggability. In addition to that flag, the patch series adds a pragma to selectively disable `-fextend-lifetimes`, and an `-fextend-this-ptr` flag which functions as `-fextend-lifetimes` for this pointers only. All changes and tests in these patches were written by @wolfy1961, though I will be managing these reviews and addressing any comments raised. The extend lifetimes flag is intended to eventually be set on by `-Og`, as discussed in the RFC [here](https://discourse.llvm.org/t/rfc-redefine-og-o1-and-add-a-new-level-of-og/72850).

---

This patch implements a new intrinsic instruction in LLVM, `llvm.fake.use` in IR and `FAKE_USE` in MIR, that takes a single operand and has no effect other than "using" its operand, to ensure that its operand remains live until after the fake use. This patch does not emit fake uses anywhere; the next patch in this sequence causes them to be emitted from the clang frontend, such that for each variable (or this) a fake.use operand is inserted at the end of that variable's scope, using that variable's value. This patch covers everything post-frontend, which is largely just the basic plumbing for a new intrinsic/instruction, along with a few steps to preserve the fake uses through optimizations (such as moving them ahead of a tail call or translating them through SROA).

wolfy1961 committed:
[ExtendLifetimes] Implement llvm.fake.use to extend variable lifetimes (#86149)

This patch is part of a set of patches that add an `-fextend-lifetimes`
flag to clang, which extends the lifetimes of local variables and
parameters for improved debuggability. In addition to that flag, the
patch series adds a pragma to selectively disable `-fextend-lifetimes`,
and an `-fextend-this-ptr` flag which functions as `-fextend-lifetimes`
for this pointers only. All changes and tests in these patches were
written by Wolfgang Pieb (@wolfy1961), while Stephen Tozer (@SLTozer)
has handled review and merging. The extend lifetimes flag is intended to
eventually be set on by `-Og`, as discussed in the RFC
here:

https://discourse.llvm.org/t/rfc-redefine-og-o1-and-add-a-new-level-of-og/72850

This patch implements a new intrinsic instruction in LLVM,
`llvm.fake.use` in IR and `FAKE_USE` in MIR, that takes a single operand
and has no effect other than "using" its operand, to ensure that its
operand remains live until after the fake use. This patch does not emit
fake uses anywhere; the next patch in this sequence causes them to be
emitted from the clang frontend, such that for each variable (or this) a
fake.use operand is inserted at the end of that variable's scope, using
that variable's value. This patch covers everything post-frontend, which
is largely just the basic plumbing for a new intrinsic/instruction,
along with a few steps to preserve the fake uses through optimizations
(such as moving them ahead of a tail call or translating them through
SROA).

Co-authored-by: Stephen Tozer <stephen.tozer@sony.com>

felipepiovezan commented:
I only skimmed the diff, but noticed there is no global isel lowering here. Were there any challenges in doing that or was it not considered?

pogo59 commented:
> I only skimmed the diff, but noticed there is no global isel lowering here. Were there any challenges in doing that or was it not considered?

Our target of interest is X86, so I believe it was not considered. @wolfy1961 can correct me if I'm mis-remembering.

SLTozer commented:
> I only skimmed the diff, but noticed there is no global isel lowering here. Were there any challenges in doing that or was it not considered?

I wasn't the original patch writer, but I think it would be the latter. I'm not familiar with the internals of GlobalISel, so I'm not sure how confident I'd be implementing support for fake use there myself - though FWIW, since it's a no-op it's mostly just filling in the boilerplate, with a few more complex cases (such as moving fake uses before tail calls).

felipepiovezan commented:
> > I only skimmed the diff, but noticed there is no global isel lowering here. Were there any challenges in doing that or was it not considered?
> 
> I wasn't the original patch writer, but I think it would be the latter. I'm not familiar with the internals of GlobalISel, so I'm not sure how confident I'd be implementing support for fake use there myself - though FWIW, since it's a no-op it's mostly just filling in the boilerplate, with a few more complex cases (such as moving fake uses before tail calls).

The reason I asked is -- and I could be wrong -- that GlobalISel will give up and fall back to SelectionDAG when it seems something it doesn't know how to lower. That usually happens in IRTranslator.cpp:

```
        if (translate(Inst))
          continue;

        ...
        reportTranslationError(*MF, *TPC, *ORE, R);
        return false;
      }
```

So, depending on how/when these intrinsics are generated, this patch could affect a few different targets.

adrian-prantl commented:
> So, depending on how/when these intrinsics are generated, this patch could affect a few different targets.

And it would introduce a situation we usually try to avoid: the presence of debug info could significantly change code generation.

wolfy1961 commented:
> I only skimmed the diff, but noticed there is no global isel lowering here. Were there any challenges in doing that or was it not considered?

IIRC, the original patch predated Global Isel, or else it was in a very early stage at the time, so it was simply not considered. 

> Our target of interest is X86, so I believe it was not considered. @wolfy1961 can correct me if I'm mis-remembering.

The patch wasn't meant to be specific to any target, although our efforts were concentrating on X86.

SLTozer commented:
> So, depending on how/when these intrinsics are generated, this patch could affect a few different targets.

Interesting; I think falling back on SelectionDAG would produce a relatively normal result (emitting `FAKE_USE`), but hard to say whether there would be any knock-on effects. In the worst case scenario we could just drop fake uses in GlobalISel - it should be safe to do so, since the intrinsics themselves are no-ops, it just means that the values will no longer be protected from the start of ISel onwards.

> And it would introduce a situation we usually try to avoid: the presence of debug info could significantly change code generation.

This intrinsic has no technical relation to debug info; the purpose is to improve debug info quality when enabled, but the codegen effects of this flag are unrelated to whether debug info is emitted or not; if you use `-fextend-lifetimes` without `-g`, you'll prevent optimizations that would have harmed debug info even when that debug info isn't present.

wolfy1961 commented:
> > So, depending on how/when these intrinsics are generated, this patch could affect a few different targets.
> 
> And it would introduce a situation we usually try to avoid: the presence of debug info could significantly change code generation.

Conceptually, though, extend-lifetimes is purely a codegen change. It could stand on its own, even though it would only be useful to improve debuggability. I think it's important to separate the 2 concepts. Debug info generation by itself does not trigger extend-lifetimes unless we specifically also request extend-lifetimes as well.

felipepiovezan commented:
> Interesting; I think falling back on SelectionDAG would produce a relatively normal result (emitting FAKE_USE),

To clarify, the entire _function_ would be skipped by GlobalISEL, not just the lowering of one instruction.

felipepiovezan commented:
> Conceptually, though, extend-lifetimes is purely a codegen change. It could stand on its own, even though it would only be useful to improve debuggability. I think it's important to separate the 2 concepts. Debug info generation by itself does not trigger extend-lifetimes unless we specifically also request extend-lifetimes as well.

Isn't the following statement contradicting the above?

> The extend lifetimes flag is intended to eventually be set on by -Og

SLTozer commented:
> To clarify, the entire _function_ would be skipped by GlobalISEL, not just the lowering of one instruction.

Ah, I misunderstood - in which case forcing it to be dropped until a complete implementation is present would be preferable.

> Isn't the following statement contradicting the above?

`-Og` is also purely a codegen change, it doesn't control whether or not debug info is emitted; in both cases, we're telling the compiler to change codegen to be better for debugging.

arsenm commented on llvm/include/llvm/Target/Target.td:
```diff
@@ -1401,6 +1401,14 @@ def FAULTING_OP : StandardPseudoInstruction {
   let isTerminator = true;
   let isBranch = true;
 }
+def FAKE_USE : StandardPseudoInstruction {
```
How is this any different than KILL?

SLTozer replied:
The only difference in the definition here is the `AsmString`, as with most of the `DBG_` classes which differ only in `AsmString` but are handled differently by MIR passes.

arsenm replied:
I mean if there's a justification for not just re-using kill for this, it needs to be clearly documented here

SLTozer replied:
Sure, I've expanded the comment.

arsenm commented:
Passing through a pseudo through globalisel should be trivial; it will just need legalization handling approximately equivalent to whatever every target does for G_IMPLICIT_DEF

OCHyams reviewed (commented):
I've looked over the code and test changes. The contents and style LGTM overall (some minor inline comments). I'm not familiar with the legalizer or X86FloatingPoint.cpp, so if someone could take a look at those parts specifically that would be great (or @arsenm have you looked at those bits already?).

Once that has been done, and once GlobalISel support has been added, I think that's everything covered by review.

I also work at Sony so I'm hesitant to Accept this once those bits have been done (I know that I am "allowed" to, but for a larger feature it feels right if others are involved too).

For added confidence, it should also be noted that the functionality has been working well for us downstream for over 5 years (possibly 7 years, but I haven't done much archeology on it) - we work at tot and we've been maintaining it since, so we know it works with today's LLVM. The feature was first mentioned upstream all the way back in 2015, here https://lists.llvm.org/pipermail/llvm-dev/2015-September/090610.html. That is all to say, as far as "new" features go this one has gone through a lot of real-world testing.

The feature this patch supports (-fextend-lifetimes, and improving the -Og pipeline to bring substantial debugging benefits) was discussed at EuroLLVM 2024 - many people seemed to support this idea and I didn't see anyone one raised any objection to it.

pogo59 commented:
There was also a [lightning talk at the 2017 US Dev Meeting](https://llvm.org/devmtg/2017-10/#lightning8). Our users have been happy with this feature for years.

jryans commented:
I hope to provide an additional review of this work today / tomorrow.

jryans reviewed (commented):
Similar to @OCHyams, this looks good to me overall, though it does touch some parts I am not familiar with (like the legalizer). Sony's multi-year usage of this downstream adds confidence in the work. The feature as a while has received positive feedback in various discussions, so it would be great to move forward here.

I believe the only major missing component is `GlobalISel` support. Once that has been added, I think this would be ready to go.

SLTozer commented on llvm/test/DebugInfo/X86/Inputs/check-fake-use.py:
```diff
@@ -0,0 +1,100 @@
+# Parsing dwarfdump's output to determine whether the location list for the
```
I agree in principle, the reason it's done here though is that the fake use can't guarantee liveness to the very end of the function, and so we can correctly lose "b" before a copy+return at the end of the function. The script adjusts for this by allowing a small numeric difference between the end of b's live range and the subprogram high PC; the way that I've updated the test to not be rotten means that _right now_ b's live range extends to the end of the function, so if that's good enough for now then I can rewrite this to use FileCheck.

dwblaikie replied:
if it works right now, and reasonably can continue to do so for the specific example being tested - I'd keep it simple/test it exactly. If there's a specific case where we can't get the whole range - maybe add a test for that and, agian, test the exact behavior with a note "this doesn't extend to the very end of the function, and that's OK or that's a known limitation, or that's a known bug (with link to the bug)"

dwblaikie commented on llvm/test/DebugInfo/X86/fake-use.ll:
```diff
@@ -0,0 +1,98 @@
+; REQUIRES: object-emission
+
+; Make sure the fake use of 'b' at the end of 'foo' causes location information for 'b'
+; to extend all the way to the end of the function.
```
Making something that pretty firmly forces optimizations that would reuse the storage shouldn't be too hard...

arsenm commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -1099,11 +1099,46 @@ void AsmPrinter::emitFunctionEntryLabel() {
 
+// Recognize cases where a spilled register is reloaded solely to feed into a
+// FAKE_USE.
+static bool isLoadFeedingIntoFakeUse(const MachineInstr &MI) {
+  const MachineFunction *MF = MI.getMF();
+  const TargetInstrInfo *TII = MF->getSubtarget().getInstrInfo();
+
```
Should probably just pass in TII instead of doing the 4 level dance to get to it

arsenm commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -1099,11 +1099,46 @@ void AsmPrinter::emitFunctionEntryLabel() {
+    return false;
+
+  unsigned Reg = MI.getOperand(0).getReg();
+  for (const MachineOperand &MO : NextI->operands()) {
+    // Return true if we came across the register from the
+    // previous spill instruction that is killed in NextI.
+    if (MO.isReg() && MO.isUse() && MO.isKill() && MO.getReg() == Reg)
+      return true;
+  }
```
This doesn't really work correctly for register aliases. Is this the same as MI.killsRegister? 

Also it would be best to avoid relying on kill flags. Hopefully we will eliminate them someday

SLTozer replied:
Do you have a suggestion for what to use instead? The intent here is to determine whether the only reason this load exists is for the immediately following fake_use; if we just removed the check for `MO.isKill()`, this would presumably lead to us deleting loads that are actually being used. Is there some alternative information, such as an analysis, that can determine whether any other instruction is using the loaded value?

arsenm replied:
The preferred method is to maintain LiveRegUnits while walking from the end of the block (which is the wrong direction for the asmprinter...)

SLTozer commented:
Apologies for the delay on updating this patch, my attention has been elsewhere for a while. I've added GlobalISel support that seems to work - in the new and updated `check-fake-use.py`, we test that we can successfully extend the lifetimes of variables for `--global-isel=1 -mtriple=aarch64--linux-gnu`.

Re: `check-fake-use.py`, I agree that it's not ideal to need a script for this. Having looked a little more at it, I've changed the script so that instead of checking that the variable's location range covers from `DW_AT_low_pc` to within 16 bytes of `DW_AT_high_pc`, we check that the variable's location range covers at least from `prologue_end`* to `epilogue_begin`. We could get rid of the python script if we currently had a measurement in LLVM for "scope bytes covered excluding prologue/epilogue", but we don't, so for now I think this needs to stay.

* It might make sense to use `DW_AT_low_pc` rather than `prologue_end` when talking about parameter lifetimes, but for the ARM GlobalISel version the prologue precedes the first `DBG_VALUE` instruction (and this is correct by IR semantics imo!) so I chose `prologue_end` in this case.

dwblaikie commented:
A prologue begin/end solution only applies to top level locals in a function, right? For smaller scopes they won't extend that far and would need to be measured relative to the high/low/ranges bytes covered by the scope the variable is within (& of course that'll always neglect the issues of variables starting their lifetime part-way through their scope (`{ foo(); int x; foo(); }` won't be live for the first foo call, etc))

Though if llvm-dwarfdump statistics need improvement, I'd hope that could be done rather than building a statistic in a separate script, by the sounds of it?

SLTozer commented:
> A prologue begin/end solution only applies to top level locals in a function, right?

Yes, the main point here is about what's needed to test this specific case more than a general point about `llvm-dwarfdump --statistics`, although there are several things that could be improved about that measurement as you've suggested. I think to summarize, there is a larger scale problem of how to measure debug scope coverage, and a smaller problem of how to determine whether -fextend-lifetimes has worked, which is conceptually simple (does the variable have a `DW_AT_location` that covers `[prologue_end, epilogue_begin)`), but for which there is no existing LLVM component that tells us that.

Though, if using a python script is a major blocker there are a few alternatives I think: we could implement this statistic in llvm-dwarfdump (though I'm not sure whether that would be a justifiable change); we could make this a Dexter test in cross-project-tests (at the cost of involving extra components that aren't really related to the test); or we could extract this information some other way, e.g. using some series of `sed` commands.

OCHyams reviewed (commented):
> Since python is already a dependency for running the tests I don't see a problem with it. While I am not sure we have tests implemented in python for clang, lldb has a lot of them so it's not unprecedented. I am interested in hearing what the objections would be in this case.

I can't speak for @dwblaikie but my objection was based on the fact that I would personally rather see a more brittle test that is simpler. It's not an objection against using python, just a question of whether it's really needed here.

That said, I don't feel strongly enough to insist and can see the benefit in having a slightly fuzzy test. Especially if it's going to be fiddly due to different debug info emission coming out of SelectionDAG/GlobalISel.

Does the script handle exprloc `DW_AT_location`s (i.e., single location that implicitly covers the scope with no explicit range)?

SLTozer commented on llvm/test/DebugInfo/X86/fake-use.ll:
```diff
@@ -0,0 +1,112 @@
+; RUN: %llc_dwarf -O2 -debugger-tune=sce --global-isel=1 -mtriple=aarch64--linux-gnu -filetype=obj -dwarf-linkage-names=Abstract < %s | llvm-dwarfdump --debug-info --debug-line -v - -o %t
+; RUN: %python %p/Inputs/check-fake-use.py %t
+; RUN: sed -e 's,call void (...) @llvm.fake.use,;,' %s | %llc_dwarf - -O2 -debugger-tune=sce -filetype=obj -dwarf-linkage-names=Abstract | llvm-dwarfdump --debug-info --debug-line -v - -o %t
+; RUN: not %python %p/Inputs/check-fake-use.py %t
```
Perhaps - it didn't occur to me because we _probably_ shouldn't see any other fail state emerge as a result of just turning off extend-lifetimes, but no harm in making it more specific!

arsenm commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -2151,6 +2151,15 @@ bool IRTranslator::translateKnownIntrinsic(const CallInst &CI, Intrinsic::ID ID,
+      for (auto VReg : getOrCreateVRegs(*Arg))
+        VRegs.push_back(VReg);
+    MIRBuilder.buildInstr(TargetOpcode::FAKE_USE, std::nullopt, VRegs,
+                          MachineInstr::copyFlagsFromInstruction(CI));
```
This won't have any flags, so no point in using copyFlagsFromInstruction

SLTozer commented:
> Does the script handle exprloc `DW_AT_location`s (i.e., single location that implicitly covers the scope with no explicit range)?

No, although I _think_ that shouldn't be possible? I think we only emit that for constants or declares.

arsenm commented on llvm/test/CodeGen/X86/fake-use-tailcall.ll:
```diff
@@ -0,0 +1,23 @@
+  notail call void (...) @llvm.fake.use(i32 %v)
+  ret void
+}
+
```
Test a case where the value produced by the call isn't used by the fake use, and is the return value?

arsenm commented on llvm/test/CodeGen/X86/fake-use-suppress-load.ll:
```diff
@@ -0,0 +1,23 @@
+}
+
+declare void @llvm.fake.use(...) #1
+
+!llvm.module.flags = !{!0}
+!llvm.ident = !{!1}
+
+!0 = !{i32 1, !"wchar_size", i32 4}
+!1 = !{!"clang version 9.0.0"}
+!2 = !{i32 -2147471544}
```
Don't need the metadata

arsenm commented on llvm/test/CodeGen/X86/fake-use-sroa.ll:
```diff
@@ -0,0 +1,54 @@
+; RUN: opt -S -passes=sroa %s | FileCheck %s
```
This should go in test/Transforms/SROA, not codegen

arsenm commented on llvm/test/CodeGen/X86/fake-use-split-ret.ll:
```diff
@@ -0,0 +1,53 @@
+; RUN: opt -mtriple=x86_64-unknown-unknown -S -codegenprepare <%s -o - | FileCheck %s
```
Move to test/Transforms/CodeGenPrepare/

arsenm commented on llvm/test/CodeGen/X86/fake-use-split-ret.ll:
```diff
@@ -0,0 +1,53 @@
+;  return bar(i);
+;}
+
+; ModuleID = 'test.cpp'
+source_filename = "test.cpp"
+target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-unknown"
```
Don't need this

arsenm commented on llvm/test/CodeGen/X86/fake-use-split-ret.ll:
```diff
@@ -0,0 +1,53 @@
+
+declare i32 @_Z3bari(i32) local_unnamed_addr
+
+; Function Attrs: nounwind
```
Drop these comments

arsenm commented on llvm/test/CodeGen/X86/fake-use-split-ret.ll:
```diff
@@ -0,0 +1,53 @@
+  ret i32 %retval.0
+}
+
+!llvm.module.flags = !{!0, !1}
+!llvm.ident = !{!2}
+
+!0 = !{i32 1, !"wchar_size", i32 2}
+!1 = !{i32 7, !"PIC Level", i32 2}
+!2 = !{!"clang version 7.0.0"}
```
Don't need the metadata

arsenm commented on llvm/test/CodeGen/X86/fake-use-hpfloat.ll:
```diff
@@ -0,0 +1,17 @@
+; assert in DAGlegalizer with fake use of half precision float.
+; Changes to half float promotion.
+; RUN: llc -O2 -stop-after=finalize-isel -filetype=asm -o - %s | FileCheck %s
```
-O2 and -filetype=asm are the defaults

arsenm commented on llvm/test/CodeGen/MIR/X86/fake-use-zero-length.ll:
```diff
@@ -0,0 +1,40 @@
+
+!0 = !{i32 1, !"wchar_size", i32 2}
+!1 = !{i32 7, !"PIC Level", i32 2}
+!2 = !{!"clang version 10.0.0"}
```
Missing newline at end of file

arsenm commented on llvm/test/CodeGen/MIR/X86/fake-use-zero-length.ll:
```diff
@@ -0,0 +1,40 @@
+; RUN: llc < %s -stop-after=finalize-isel | FileCheck %s --implicit-check-not=FAKE_USE
```
test/CodeGen/MIR is for testing of the MIR printer/parser, not tests that check MIR output. This should go to test/CodeGen/X86

arsenm commented on llvm/test/CodeGen/MIR/X86/fake-use-phi.mir:
```diff
@@ -0,0 +1,95 @@
+# RUN: llc < %s -x mir -run-pass=codegenprepare | FileCheck %s --implicit-check-not="llvm.fake.use"
```
This test doesn't make sense, and I'm surprised this doesn't error on you. You're trying to run an IR pass on MIR, and the IR is supposed to just be a constant here and the MIR should not exist yet. You should move this to test/CodeGenPrepare and run it with opt

OCHyams commented:
>> I said
>> Does the script handle exprloc DW_AT_locations (i.e., single location that implicitly covers the scope with no explicit range)?
>
>@SLTozer said
> No, although I think that shouldn't be possible? I think we only emit that for constants or declares.

It's definitely possible:
```
$ cat test.ll
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-unknown"

define i32 @f(i32 %a) !dbg !4 {
entry:
  tail call void @llvm.dbg.value(metadata i32 %a, metadata !8, metadata !DIExpression()), !dbg !10
  ret i32 0, !dbg !10
}

declare void @llvm.dbg.value(metadata, metadata, metadata)

!llvm.dbg.cu = !{!0}
!llvm.debugify = !{!2, !2}
!llvm.module.flags = !{!3}

!0 = distinct !DICompileUnit(language: DW_LANG_C, file: !1, producer: "debugify", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug)
!1 = !DIFile(filename: "test.ll", directory: "/")
!2 = !{i32 1}
!3 = !{i32 2, !"Debug Info Version", i32 3}
!4 = distinct !DISubprogram(name: "f", linkageName: "f", scope: null, file: !1, line: 1, type: !5, scopeLine: 1, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, retainedNodes: !7)
!5 = !DISubroutineType(types: !6)
!6 = !{}
!7 = !{!8}
!8 = !DILocalVariable(name: "1", scope: !4, file: !1, line: 1, type: !9)
!9 = !DIBasicType(name: "ty32", size: 32, encoding: DW_ATE_unsigned)
!10 = !DILocation(line: 1, column: 1, scope: !4)
```
```
$ llc test2.ll --filetype=obj -o - | llvm-dwarfdump -
<...>
0x00000043:     DW_TAG_variable
                  DW_AT_location        (DW_OP_reg5 RDI)
                  DW_AT_name    ("1")
                  DW_AT_decl_file       ("/test.ll")
                  DW_AT_decl_line       (1)
                  DW_AT_type    (0x00000051 "ty32")
 ```

SLTozer commented:
> It's definitely possible:

Ah that's true, if it's going to be live at a single location for every instruction in the function then it'll emit with a fixed location. That shouldn't happen in this test since we have the big register kill operation, but no point in making assumptions when we can add coverage instead - will update.

arsenm commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -1099,11 +1099,45 @@ void AsmPrinter::emitFunctionEntryLabel() {
   const MachineFunction *MF = MI.getMF();
   const TargetInstrInfo *TII = MF->getSubtarget().getInstrInfo();
 
+  // If this is a reload of a spilled register that only feeds into a FAKE_USE
+  // instruction, meaning the load value has no effect on the program and has
+  // only been kept alive for debugging; since it is still available on the
+  // stack, we can skip the load itself.
```
I don't understand the comment in this context. The load isn't being skipped, it's just skipping the comments about the instruction

arsenm commented on llvm/test/CodeGen/MIR/X86/fake-use-scheduler.mir:
```diff
@@ -0,0 +1,223 @@
+  attributes #1 = { nocallback nofree nosync nounwind willreturn }
+  
+  !llvm.module.flags = !{!0}
+  !llvm.ident = !{!1}
```
Can drop this

arsenm commented on llvm/test/CodeGen/MIR/X86/fake-use-scheduler.mir:
```diff
@@ -0,0 +1,223 @@
+    %0:gr64 = COPY $rdi
+    %1:gr64_with_sub_8bit = MOVSX64rm32 $rip, 1, $noreg, @glb, $noreg :: (dereferenceable load (s32) from @glb, align 16, !tbaa !2)
+    MOV32mr %0, 1, $noreg, 0, $noreg, %1.sub_32bit :: (store (s32) into %ir.p, !tbaa !2)
+    %3:gr64_with_sub_8bit = MOVSX64rm32 $rip, 1, $noreg, @glb + 4, $noreg :: (dereferenceable load (s32) from `ptr getelementptr inbounds ([100 x i32], ptr @glb, i64 0, i64 1)`, !tbaa !2)
```
Drop the IR references?

arsenm commented on llvm/test/CodeGen/MIR/X86/fake-use-scheduler.mir:
```diff
@@ -0,0 +1,223 @@
+liveins:
+  - { reg: '$rdi', virtual-reg: '%0' }
+frameInfo:
+  isFrameAddressTaken: false
```
Can remove most of this, -simplify-mir helps

arsenm commented on llvm/test/CodeGen/MIR/X86/fake-use-scheduler.mir:
```diff
@@ -0,0 +1,223 @@
+# Prevent the machine scheduler from moving instructions past FAKE_USE.
+# RUN: llc -run-pass machine-scheduler -debug-only=machine-scheduler 2>&1 -o - %s | FileCheck %s
```
This belongs in test/CodeGen/X86, not under MIR as it is not a MIR support test

arsenm commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -1843,6 +1880,12 @@ void AsmPrinter::emitFunctionBody() {
+        // FAKE_USE instruction, meaning the load value has no effect on the
+        // program and has only been kept alive for debugging; since it is
+        // still available on the stack, we can skip the load itself.
+        if (isLoadFeedingIntoFakeUse(MI, TII))
+          break;
```
It's pretty inappropriate for an instruction to disappear in the AsmPrinter. Can you handle this optimization the same way we do for kill?

dwblaikie commented:
> > Since python is already a dependency for running the tests I don't see a problem with it. While I am not sure we have tests implemented in python for clang, lldb has a lot of them so it's not unprecedented. I am interested in hearing what the objections would be in this case.
> 
> I can't speak for @dwblaikie but my objection was based on the fact that I would personally rather see a more brittle test that is simpler. It's not an objection against using python, just a question of whether it's really needed here.

Yeah, I jumped in the middle a bit - I haven't looked at the python script/tests in question yet.

But perhaps one example to discuss would be useful? 

& yeah, for sufficiently narrow/single-transform tests I don't think it'd be that bad to test for the exact result...

SLTozer commented:
Thank you @arsenm for the thorough reviews - I'll fix up the remaining comments (regarding AsmPrinter) early tomorrow.

> But perhaps one example to discuss would be useful?

The script only applies to one test right now, though it's run in 4 different ways; the script is specialized for the test file, but it could be parameterized. The purpose of this test is essentially checking that FAKE_USE does what it says on the tin, that a variable that should otherwise by all rights be killed is available for its full scope.

The difficult part is that the test still isn't _narrow_; the test runs `llc` from IR to object with optimizations enabled and checks at the end that the variable under test is still available for its full scope. We could try making the checks exact, but it'd mean 4 different checks (or more if we increase coverage further) that could each change if any pass changes its behaviour. The actual check is simple I think - we read the output of dwarfdump to confirm that the parameter has a continuous location range that extends from `[prologue_end, epilogue_begin)`. Ideally we'd have a proper way of doing this, but I think it'd be messy and brittle to substitute exact checks.

I think I'd be cautious on principle about adding special scripts to run a single test, but I don't see any easy alternatives without major drawbacks, and I think this is worth testing for; ultimately if the verdict comes down against the python script, my backup proposal is to make it a Dexter test (which will need to be moved to the subsequent patch which adds front-end support).

arsenm commented on llvm/lib/CodeGen/RemoveLoadsIntoFakeUses.cpp:
```diff
@@ -0,0 +1,170 @@
+
+  SmallDenseMap<Register, SmallVector<MachineInstr *>> RegFakeUses;
+  LivePhysRegs.init(*TRI);
+  dbgs() << "Running the pass!\n";
```
Stray debug printing

SLTozer replied:
I've pushed up the removal for those already, but this was some very quick reviewing!

arsenm commented on llvm/lib/CodeGen/RemoveLoadsIntoFakeUses.cpp:
```diff
@@ -0,0 +1,170 @@
+      skipFunction(MF.getFunction()))
+    return false;
+
+  dbgs() << "We could run the pass!\n";
```
Stray debug printing

arsenm commented on llvm/lib/CodeGen/RemoveLoadsIntoFakeUses.cpp:
```diff
@@ -0,0 +1,170 @@
+        if (!LivePhysRegs.available(Reg) || MRI->isReserved(Reg)) {
+          dbgs() << "Reg is live: ";
+          TRI->dumpReg(Reg);
+          dbgs() << "\n";
```
More leftover debugging

arsenm commented on llvm/lib/CodeGen/RemoveLoadsIntoFakeUses.cpp:
```diff
@@ -0,0 +1,170 @@
+INITIALIZE_PASS_END(RemoveLoadsIntoFakeUses, DEBUG_TYPE,
+                    "Remove Loads Into Fake Uses", false, false)
+
+bool RemoveLoadsIntoFakeUses::runOnMachineFunction(MachineFunction &MF) {
```
I wonder if we can just do this in ExpandPostRAPseudos and save having yet another pass.

This would also bring the side effect where I think we should have ExpandPostRAPseudos run in reverse and lazily maintain LiveRegUnits

SLTozer replied:
I'm fine to move it into another pass, and patch up that pass as I go through it - just a quick question though: Might this step too far outside the scope of what `ExpandPostRAPseudos` is expected to do? If so, would that justify changing the name slightly, or do we prefer to keep those fixed over time?

And also a minor unrelated, I noticed that passes may be added to `TargetPassConfig::addMachinePasses` or to `CodeGenPassBuilder<Derived>::addMachinePasses`, and most (but not all) passes are in both - is there an important distinction between them for determining where a new pass should go?

arsenm replied:
I do believe we have an elide-reload-of-kill optimization but a 1 minute scan didn't find it, wherever that happens would be the best place.
TargetPassConfig::addMachinePasses = OldPM CodeGenPassBuilder = NewPM and incomplete

SLTozer replied:
Might be `MachineLateInstrsCleanup`? I see that that relies on `isKill()` for liveness checks as well, but might otherwise make sense as a location. It's hard to tell whether there are any optimizations afterwards that might cause an issue; it'd probably be faster to implement and test it than scan through the source of all the applicable passes.

arsenm replied:
It's not MachineLateInstrsCleanup  (which is also running in the wrong direction). It's looking for killed register operands, not KILL instructions

SLTozer commented:
It ended up taking a little longer than I expected to make the AsmPrinter changes; I couldn't see a sensible place to put the "remove loads feeding into fake use" behaviour, so I made it into a pass. The main goal here was to have this take place very late, so that no subsequent pass takes advantage of the removed load/fake_use to shorten the stack lifetime of a variable, but before LiveDebugValues, so that LDV has an accurate understanding of which machine locations house the required values.

The pass itself runs only if the function has `optdebug` enabled, since those should be the only functions with fake uses (tests have been updated to reflect this); it's not an error if we don't run this pass for a function that has fake use instructions though. I'm happy to move it to a separate patch if it makes it easier for reviewers - even outside of test files, this patch is already very large.

OCHyams commented:
> Ah that's true, if it's going to be live at a single location for every instruction in the function then it'll emit with a fixed location. That shouldn't happen in this test since we have the big register kill operation, but no point in making assumptions when we can add coverage instead - will update.

Ah ok if it can't occur in the test then no need. We can extend it as and when necessary (that's better than having additional complexity that isn't used in the script).

igorkudrin commented:
[test/CodeGen/X86/fake-use-escape.ll](https://reviews.llvm.org/D41043#change-rq7Ra9xJmIVE) from the original patch triggers an assertion failure `isAllocaPromotable(AI) && "Cannot promote non-promotable alloca!"`, file `llvm\lib\Transforms\Utils\PromoteMemoryToRegister.cpp:737`.

SLTozer commented:
> [test/CodeGen/X86/fake-use-escape.ll](https://reviews.llvm.org/D41043#change-rq7Ra9xJmIVE) from the original patch triggers an assertion failure `isAllocaPromotable(AI) && "Cannot promote non-promotable alloca!"`, file `llvm\lib\Transforms\Utils\PromoteMemoryToRegister.cpp:737`.

I see - the reason I removed that test from the patch was that the circumstances of the crash could only be reproduced with typed pointers, which are no longer supported in LLVM. Likewise, the code that the test was covering has been removed, because as far as I can tell there is no way to trigger it without typed pointers. While it's unclear whether we'd need to support this in textual IR, since we don't guarantee backwards compatibility, the error also occurs for the bitcode version of the file which we do expect to keep supporting over time, so I'll restore this test.

arsenm commented on llvm/lib/CodeGen/RemoveLoadsIntoFakeUses.cpp:
```diff
@@ -0,0 +1,166 @@
+    return false;
+
+  // This implementation assumes we are post-RA.
+  if (!MF.getProperties().hasProperty(
+          MachineFunctionProperties::Property::NoVRegs))
```
Instead of this check, should implement getRequiredProperties and let the verifier check it

arsenm reviewed (approved):
lgtm with nit. I think some more passes will need updating to deal with these (like InferAddressSpaces) but it's easiest to do that incrementally

SLTozer commented:
> lgtm with nit. I think some more passes will need updating to deal with these (like InferAddressSpaces) but it's easiest to do that incrementally

Thanks very much for the amount of work put into this review - I'll rebase now and highlight if there are any notable conflicts/changes before merging.

SLTozer commented:
Rebased onto current main; the only substantive change made is that RemoveLoadsIntoFakeUses is explicitly disabled for WASM, SPIRV, and NVPTX, in exactly the same way as all other passes that require NoVRegs.

SLTozer commented:
Looks like an explicit triple is needed for some of the tests (should probably go on all of them, but for now I'll just add it to the tests that are failing).

tobias-stadler commented:
Is there a good reason for not putting isFakeUse() into MachineInstr::isPosition() instead of sprinkling it over the codebase?
Putting into isPosition would make it a scheduling boundary, ignored by CSE, unsafe to move and therefore ignored by DCE.

SLTozer commented:
> Sorry for only raising this now. Is there a good reason for not putting isFakeUse() into MachineInstr::isPosition() instead of sprinkling it over the codebase? Putting into isPosition would make it a scheduling boundary, ignored by CSE, unsafe to move and therefore ignored by DCE and MachineSink.

It's not something I've thought about, but I'm happy to take a look at it and see if a refactoring patch would be worthwhile.

tobias-stadler commented:
> It's not something I've thought about, but I'm happy to take a look at it and see if a refactoring patch would be worthwhile.

Thanks! Just to clarify, I am not 100% sure that isPosition is the correct place to put this, but it seems like the best fit. If we find that isPosition doesn't work, we probably still want FAKE_USE to return false in MachineInstr::isSafeToMove() in some way or another though. isDebugInstr might also be worth considering, but I'm guessing it's not really a debug instruction?

pogo59 commented:
> isDebugInstr might also be worth considering, but I'm guessing it's not really a debug instruction?

FAKE_USE influences code generation so as a matter of principle it should not be thought of as a debug instruction. (Those are going away anyhow, but there's a dividing line there.)

Prabhuk commented:
We are seeing a test failure in our builders which I think is related to this patch:

https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-base-mac-x64/b8738251281152890817/overview

```
FAIL: LLVM :: DebugInfo/X86/fake-use.ll (28733 of 54710)
******************** TEST 'LLVM :: DebugInfo/X86/fake-use.ll' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 6: /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llc -O2 -filetype=obj -dwarf-linkage-names=Abstract < /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/DebugInfo/X86/fake-use.ll | /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llvm-dwarfdump --debug-info --debug-line -v - -o /Volumes/Work/s/w/ir/x/w/llvm_build/test/DebugInfo/X86/Output/fake-use.ll.tmp
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llc -O2 -filetype=obj -dwarf-linkage-names=Abstract
+ /Volumes/Work/s/w/ir/x/w/llvm_build/bin/llvm-dwarfdump --debug-info --debug-line -v - -o /Volumes/Work/s/w/ir/x/w/llvm_build/test/DebugInfo/X86/Output/fake-use.ll.tmp
RUN: at line 7: "/Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/bin/python3.9" /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/DebugInfo/X86/../Inputs/check-fake-use.py /Volumes/Work/s/w/ir/x/w/llvm_build/test/DebugInfo/X86/Output/fake-use.ll.tmp
+ /Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/bin/python3.9 /Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/DebugInfo/X86/../Inputs/check-fake-use.py /Volumes/Work/s/w/ir/x/w/llvm_build/test/DebugInfo/X86/Output/fake-use.ll.tmp
Traceback (most recent call last):
  File "/Volumes/Work/s/w/ir/x/w/llvm-llvm-project/llvm/test/DebugInfo/X86/../Inputs/check-fake-use.py", line 57, in <module>
    raise RuntimeError(f"Invalid location range list for 'b'")
RuntimeError: Invalid location range list for 'b'
```

I am investigating further to confirm.

SLTozer commented:
> I am investigating further to confirm.

Some of the tests added in this patch didn't have an explicit triple, which was necessary for the test to pass - I've already committed updates to add those triples, by the looks of it the test has started passing as of that commit:
https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-base-mac-x64/b8738234501894039297/overview
